packages/contracts/utils/storage.ts

```diff
@@ -0,0 +1,26 @@
+import {ethers} from 'hardhat';
+import {defaultAbiCoder} from 'ethers/lib/utils';
+
+import {IMPLEMENTATION_SLOT} from '../test/test-utils/uups-upgradeable';
+
+export async function readImplementationValuesFromSlot(
+  contractAddresses: string[]
+): Promise<string[]> {
+  const implementationValues: string[] = await Promise.all(
+    contractAddresses.map(async contractAddress => {
+      return readImplementationValueFromSlot(contractAddress);
+    })
+  );
+
+  return implementationValues;
+}
+
+export async function readImplementationValueFromSlot(
+  contractAddress: string
+): Promise<string> {
+  const encoded = await ethers.provider.getStorageAt(
+    contractAddress,
+    IMPLEMENTATION_SLOT
+  );
+  return defaultAbiCoder.decode(['address'], encoded)[0];
```

Comment:
Same as before, simple promises are lighter to run

```suggestion
export function readImplementationValueFromSlot(
  contractAddress: string
): Promise<string> {
  return ethers.provider.getStorageAt(
    contractAddress,
    IMPLEMENTATION_SLOT
  ).then((encoded) => 
    defaultAbiCoder.decode(['address'], encoded)[0]
  );
```